### PR TITLE
Add ExternalConsultant & basic logic for it

### PR DESCRIPTION
**Pull request by rnudb**

### CODE REVIEW

File: interview/views.py [hunk 1]
```diff
@@ -415,8 +417,12 @@ def new_candidate(request, past_candidate_id=None):
         candidate_form = ProcessCandidateForm()
         process_form = ProcessForm()
         interviewers_form = InterviewersForm(prefix="interviewers")
-        process_form.fields["subsidiary"].initial = request.user.consultant.company.id
 
+        if request.user.consultant.company:
```
Comment (rnudb) [inline interview/views.py:421]: needs changes: an external consultant is always linked to a subsidiary

File: ref/models.py [hunk 2]
```diff
@@ -111,6 +111,9 @@ class Consultant(models.Model):
     user = models.OneToOneField(settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
     company = models.ForeignKey(Subsidiary, verbose_name=_("Subsidiary"), null=True, on_delete=models.SET_NULL)
 
+    # dst class written with string to avoid circular imports issues
+    to_source = models.ForeignKey("interview.Sources", null=True, default=None, on_delete=models.SET_NULL)
```
Comment (rnudb) [inline ref/models.py:115]: add blank=True

**Review by bport — CHANGES_REQUESTED** [review]

File: interview/views.py [hunk 3]
```diff
@@ -664,6 +682,9 @@ def edit_candidate(request, process_id):
     source_form = SourceForm(prefix="source")
     offer_form = OfferForm(prefix="offer")
 
+    if request.user.consultant.to_source:
```
[the remark this answers is not in the document]
Comment (rnudb) [reply interview/views.py:685]: External user can and will reach this line. If we remove it, the source field will appear in the form which is not what's wanted (source predefined => no source field).

File: interview/views.py [hunk 4]
```diff
@@ -417,6 +422,8 @@ def new_candidate(request, past_candidate_id=None):
         interviewers_form = InterviewersForm(prefix="interviewers")
         process_form.fields["subsidiary"].initial = request.user.consultant.company.id
 
+    if request.user.consultant.to_source:
```
[the remark this answers is not in the document]
Comment (rnudb) [reply interview/views.py:425]: Same as above comment

[the remark this answers is not in the document]
Comment (bport) [reply interview/views.py:425]: my bad

File: ref/models.py [hunk 5]
```diff
@@ -105,12 +105,34 @@ def create_consultant(self, trigramme, email, company, full_name, **extra_fields
...
 
 
+def is_not_external_check(user):
+    return user.consultant.limited_to_source is None
+
+
 class Consultant(models.Model):
     """A consultant that can do recruitment meeting"""
 
+    ALL = 1
```
Comment (bport) [inline ref/models.py:115]: Use IntegerChoice enum style

https://docs.djangoproject.com/en/3.2/ref/models/fields/#enumeration-types

File: ref/models.py [hunk 6]
```diff
@@ -105,12 +105,34 @@ def create_consultant(self, trigramme, email, company, full_name, **extra_fields
...
+        (EXTERNAL_WRITE, _("User is an external consultant")),
+        (EXTERNAL_READONLY, _("User is external and has only read rights")),
+    )
+
     user = models.OneToOneField(settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
     company = models.ForeignKey(Subsidiary, verbose_name=_("Subsidiary"), null=True, on_delete=models.SET_NULL)
 
+    # dst class written with string to avoid circular imports issues
+    limited_to_source = models.ForeignKey(
+        "interview.Sources", null=True, blank=True, default=None, on_delete=models.DO_NOTHING
```
Comment (bport) [inline ref/models.py:130]: I don't think we want DO_NOTHING, this will raise a constraint error

File: ref/models.py [hunk 7]
```diff
@@ -105,12 +105,34 @@ def create_consultant(self, trigramme, email, company, full_name, **extra_fields
...
+    )
+    privilege = models.PositiveSmallIntegerField(
+        choices=PRIVILEGE_LEVEL, verbose_name=_("Authority level"), default=ALL
```
Comment (bport) [inline ref/models.py:133]: We need to translate it

**Review by bport — CHANGES_REQUESTED** [review]
Add some test to ensure all work as expected (don't need to have a test for each view)

But at least :
Test external user access on a page with right, on a source page not allow, external without source associated to him.

File: interview/models.py [hunk 8]
```diff
@@ -209,7 +209,10 @@ def __str__(self):
 
 class ProcessManager(models.Manager):
     def for_user(self, user):
-        return super().get_queryset().filter(start_date__gte=user.date_joined)
+        q = super().get_queryset().filter(start_date__gte=user.date_joined)
+        if user.consultant.privilege > Consultant.PrivilegeLevel.ALL:
```
Comment (ArnaudChirat) [inline interview/models.py:213]: Add a property to the model to avoid change every where in the code if the logic change

Comment (ArnaudChirat) [reply interview/models.py:213]: What is the difference between > Consultant.PrivilegeLevel.ALL and is_external ?

File: interview/views.py [hunk 9]
```diff
@@ -294,6 +297,12 @@ def closed_processes(request):
 @login_required
 @require_http_methods(["GET"])
 def processes_for_source(request, source_id):
+    if request.user.consultant.privilege > Consultant.PrivilegeLevel.ALL and (  # is consultant external ?
```
Comment (ArnaudChirat) [inline interview/views.py:300]: Why not using the decorator is_external ?

**Review by ArnaudChirat — CHANGES_REQUESTED** [review]
Add middleware to check if user is valid.
- If external is correct

File: interview/tests.py [hunk 10]
```diff
@@ -923,3 +925,58 @@ def test_interview_minute_edit_display(self):
         self.assertContains(response, "NO")
         self.assertContains(response, "DRAFT")
         self.assertContains(response, "GO")
+
+
```
Comment (ArnaudChirat) [inline interview/tests.py:929]: Add a test for the middleware

**Review by ArnaudChirat — CHANGES_REQUESTED** [review]

File: interview/tests.py (hunk 10 above)
Comment (rnudb) [reply interview/tests.py:930]: Done !

**Review by ArnaudChirat — APPROVED** [review]

**Pull request merged**